### PR TITLE
Add an AWS Kinesis connector to the exporting engine

### LINKED ISSUE #7169: Implement AWS Kinesis exporting connector
##### Feature idea summary

* Code already exists in the backend that can export AWS Kinesis.
* Code currently has poor structure - intertwined with the existing backends.
* Sits on top of AWS SDK in C++, unknown if this wrapper will change.
* Similar in complexity to the Prometheus remote write exporter.
* Unit-tests will need to call the wrapped C++ code - unclear if this is difficult.

##### Expected behavior

* Reviewer will require a working AWS Kinesis service to test against.
* @cakrit will need to provide access to AWS for testing properly.
* Note for the reviewer - the threading/locking should be standard for the exporting engine, may need to read the config reading code.

### PR DESCRIPTION
**Pull request by vlvkobal**
##### Summary
AWS Kinesis exporting connector was implemented as a part of the backends refactoring project #6762.

Closes #7169

##### Component Name
backends, exporting engine

##### Description of testing that the developer performed
To build Netdata with the AWS Kinesis support use the `--enable-backend-kinesis` option.
I tested the connector with the AWS Kinesis service:
image
My `exporting.conf`:
```
[exporting:global]
enabled = yes
send configured labels = yes
send automatic labels = no

... <other connector instances>

[kinesis:instance4]
enabled = yes
destination = us-east-1
stream name = netdata
aws_access_key_id = <id>
aws_secret_access_key = <secret>
data source = as collected
update every = 2
send charts matching = system.processes

[kinesis:instance5]
enabled = yes
destination = us-east-1
stream name = second
aws_access_key_id = <id>
aws_secret_access_key = <secret>
data source = as collected
update every = 2
send charts matching = system.processes

```

To get data from the database I used the AWS CLI with the `config` and `credentials` configuration files in the `~/.aws` directory:
```
[default]
region=us-east-1
output=json
```
```
[default]
aws_access_key_id=<id>
aws_secret_access_key=<secret>

```

CLI commands:
```
aws kinesis list-shards --stream-name netdata
aws kinesis get-shard-iterator --stream-name netdata --shard-id "shardId-000000000000" --shard-iterator-type LATEST
aws kinesis get-records --shard-iterator=<iterator from the previous step>
```
A correct answer contains something like that:
```
{
            "SequenceNumber": "49604358198945158955635420576740905361319834852404494338",
            "ApproximateArrivalTimestamp": 1582216988.906,
            "Data": "eyJwcmVmaXgiOiJuZXRkYXRhIiwiaG9zdG5hbWUiOiJhcmNoLWVzeGkiLCJob3N0X3RhZ3MiOiJob3N0X3RhZzE9Imhvc3RfdmFsdWUxIixob3N0X3RhZzI9Imhvc3RfdmFsdWUyIiIsImxhYmVscyI6eyJob3N0X3RhZzEiOiJob3N0X3ZhbHVlMSIsImhvc3RfdGFnMiI6Imhvc3RfdmFsdWUyIiwiY29uZmlndXJlZF9ob3N0X2xhYmVsMiI6Imhvc3RfbGFiZTJfdmFsdWUiLCJjb25maWd1cmVkX2hvc3RfbGFiZWwxIjoiaG9zdF9sYWJlbF92YWx1ZSJ9LCJjaGFydF9pZCI6InN5c3RlbS5wcm9jZXNzZXMiLCJjaGFydF9uYW1lIjoic3lzdGVtLnByb2Nlc3NlcyIsImNoYXJ0X2ZhbWlseSI6InByb2Nlc3NlcyIsImNoYXJ0X2NvbnRleHQiOiJzeXN0ZW0ucHJvY2Vzc2VzIiwiY2hhcnRfdHlwZSI6InN5c3RlbSIsInVuaXRzIjoicHJvY2Vzc2VzIiwiaWQiOiJydW5uaW5nIiwibmFtZSI6InJ1bm5pbmciLCJ2YWx1ZSI6NiwidGltZXN0YW1wIjoxNTgyMjE2OTEwfQp7InByZWZpeCI6Im5ldGRhdGEiLCJob3N0bmFtZSI6ImFyY2gtZXN4aSIsImhvc3RfdGFncyI6Imhvc3RfdGFnMT0iaG9zdF92YWx1ZTEiLGhvc3RfdGFnMj0iaG9zdF92YWx1ZTIiIiwibGFiZWxzIjp7Imhvc3RfdGFnMSI6Imhvc3RfdmFsdWUxIiwiaG9zdF90YWcyIjoiaG9zdF92YWx1ZTIiLCJjb25maWd1cmVkX2hvc3RfbGFiZWwyIjoiaG9zdF9sYWJlMl92YWx1ZSIsImNvbmZpZ3VyZWRfaG9zdF9sYWJlbDEiOiJob3N0X2xhYmVsX3ZhbHVlIn0sImNoYXJ0X2lkIjoic3lzdGVtLnByb2Nlc3NlcyIsImNoYXJ0X25hbWUiOiJzeXN0ZW0ucHJvY2Vzc2VzIiwiY2hhcnRfZmFtaWx5IjoicHJvY2Vzc2VzIiwiY2hhcnRfY29udGV4dCI6InN5c3RlbS5wcm9jZXNzZXMiLCJjaGFydF90eXBlIjoic3lzdGVtIiwidW5pdHMiOiJwcm9jZXNzZXMiLCJpZCI6ImJsb2NrZWQiLCJuYW1lIjoiYmxvY2tlZCIsInZhbHVlIjowLCJ0aW1lc3RhbXAiOjE1ODIyMTY5MTB9CnsicHJlZml4IjoibmV0ZGF0YSIsImhvc3RuYW1lIjoiYXJjaC1lc3hpIiwibGFiZWxzIjp7ImxhYmVsMSI6InZhbHVlMSJ9LCJjaGFydF9pZCI6InN5c3RlbS5wcm9jZXNzZXMiLCJjaGFydF9uYW1lIjoic3lzdGVtLnByb2Nlc3NlcyIsImNoYXJ0X2ZhbWlseSI6InByb2Nlc3NlcyIsImNoYXJ0X2NvbnRleHQiOiJzeXN0ZW0ucHJvY2Vzc2VzIiwiY2hhcnRfdHlwZSI6InN5c3RlbSIsInVuaXRzIjoicHJvY2Vzc2VzIiwiaWQiOiJydW5uaW5nIiwibmFtZSI6InJ1bm5pbmciLCJ2YWx1ZSI6NSwidGltZXN0YW1wIjoxNTgyMjE2OTEwfQp7InByZWZpeCI6Im5ldGRhdGEiLCJob3N0bmFtZSI6ImFyY2gtZXN4aSIsImxhYmVscyI6eyJsYWJlbDEiOiJ2YWx1ZTEifSwiY2hhcnRfaWQiOiJzeXN0ZW0ucHJvY2Vzc2VzIiwiY2hhcnRfbmFtZSI6InN5c3RlbS5wcm9jZXNzZXMiLCJjaGFydF9mYW1pbHkiOiJwcm9jZXNzZXMiLCJjaGFydF9jb250ZXh0Ijoic3lzdGVtLnByb2Nlc3NlcyIsImNoYXJ0X3R5cGUiOiJzeXN0ZW0iLCJ1bml0cyI6InByb2Nlc3NlcyIsImlkIjoiYmxvY2tlZCIsIm5hbWUiOiJibG9ja2VkIiwidmFsdWUiOjAsInRpbWVzdGFtcCI6MTU4MjIxNjkxMH0K",
            "PartitionKey": "netdata_0"
        }
```

You can decode the data using any Base64 decoder:
```
{"prefix":"netdata","hostname":"arch-esxi","host_tags":"host_tag1="host_value1",host_tag2="host_value2"","labels":{"host_tag1":"host_value1","host_tag2":"host_value2","configured_host_label2":"host_labe2_value","configured_host_label1":"host_label_value"},"chart_id":"system.processes","chart_name":"system.processes","chart_family":"processes","chart_context":"system.processes","chart_type":"system","units":"processes","id":"running","name":"running","value":6,"timestamp":1582216910}
{"prefix":"netdata","hostname":"arch-esxi","host_tags":"host_tag1="host_value1",host_tag2="host_value2"","labels":{"host_tag1":"host_value1","host_tag2":"host_value2","configured_host_label2":"host_labe2_value","configured_host_label1":"host_label_value"},"chart_id":"system.processes","chart_name":"system.processes","chart_family":"processes","chart_context":"system.processes","chart_type":"system","units":"processes","id":"blocked","name":"blocked","value":0,"timestamp":1582216910}
{"prefix":"netdata","hostname":"arch-esxi","labels":{"label1":"value1"},"chart_id":"system.processes","chart_name":"system.processes","chart_family":"processes","chart_context":"system.processes","chart_type":"system","units":"processes","id":"running","name":"running","value":5,"timestamp":1582216910}
{"prefix":"netdata","hostname":"arch-esxi","labels":{"label1":"value1"},"chart_id":"system.processes","chart_name":"system.processes","chart_family":"processes","chart_context":"system.processes","chart_type":"system","units":"processes","id":"blocked","name":"blocked","value":0,"timestamp":1582216910}
```

##### Additional information
During the process of writing the connector, it turned out that the whole layer of the `connector` structures should be deleted. It was too complicated to do it in a separate PR so I've done it in a separate commit.

### CODE REVIEW

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/aws_kinesis/aws_kinesis_put_record.cc  2
- backends/aws_kinesis/aws_kinesis_put_record.cc  2
- exporting/tests/test_exporting_engine.c  8
- backends/backends.c  2
- exporting/json/json.c  1
- exporting/aws_kinesis/aws_kinesis.c  3
- exporting/opentsdb/opentsdb.c  2
         

Clones removed
==============
+ exporting/process_data.c  -16
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5042459&bid=16607224)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/json/json.c  1
- backends/aws_kinesis/aws_kinesis_put_record.cc  2
- exporting/aws_kinesis/aws_kinesis.c  3
- exporting/aws_kinesis/aws_kinesis_put_record.cc  2
- exporting/tests/test_exporting_engine.c  8
- backends/backends.c  2
- exporting/opentsdb/opentsdb.c  2
         

Clones removed
==============
+ exporting/process_data.c  -16
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5042459&bid=16607224)

File: exporting/aws_kinesis/aws_kinesis_put_record.cc [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    Kinesis::KinesisClient *client;
+
+    if (access_key_id && *access_key_id && secret_key && *secret_key) {
+        client = New<Kinesis::KinesisClient>("client", Auth::AWSCredentials(access_key_id, secret_key), config);
+    } else {
+        client = New<Kinesis::KinesisClient>("client", config);
+    }
+    kinesis_specific_data->client = (void *)client;
+
+    Vector<request_outcome> *request_outcomes;
```
Comment (thiagoftsm) [inline exporting/aws_kinesis/aws_kinesis_put_record.cc]: Considering that we are dealing with classes here, should we use `try .. catch`?

File: exporting/read_config.c [hunk 2]
```diff
@@ -260,107 +255,118 @@ struct engine *read_exporting_config()
...
-            tmp_connector->config.type = i;
-            tmp_connector->engine = engine;
+#ifndef HAVE_KINESIS
+        if (tmp_ci_list->backend_type == BACKEND_TYPE_KINESIS) {
+            error("AWS Kinesis support isn't compiled");
+            goto next_connector_instance;
+        }
+#endif
+
+        tmp_instance = (struct instance *)calloc(1, sizeof(struct instance));
```
Comment (thiagoftsm) [inline exporting/read_config.c]: Please, use `callocz` here instead `calloc`.

File: exporting/tests/exporting_doubles.c [hunk 3]
```diff
@@ -16,13 +16,11 @@ struct engine *__mock_read_exporting_config()
...
     engine->config.update_every = 3;
 
-    engine->connector_root = calloc(1, sizeof(struct connector));
-    engine->connector_root->config.type = BACKEND_TYPE_GRAPHITE;
-    engine->connector_root->engine = engine;
 
-    engine->connector_root->instance_root = calloc(1, sizeof(struct instance));
-    struct instance *instance = engine->connector_root->instance_root;
-    instance->connector = engine->connector_root;
+    engine->instance_root = calloc(1, sizeof(struct instance));
```
Comment (thiagoftsm) [inline exporting/tests/exporting_doubles.c]: Considering that this is a test, we do not need the `callocz`, but I believe we need to test the result.

File: exporting/aws_kinesis/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# Using Netdata with AWS Kinesis Data Streams
+
+## Prerequisites
+
+To use AWS Kinesis for metric collecting and processing, AWS SDK for C++ should be
```
Comment (shortpatti) [inline exporting/aws_kinesis/README.md]: maybe change the last part of the sentence to, "...you must [install](https://docs.aws.amazon.com/en_us/sdk-for-cpp/v1/developer-guide/setup.html) AWS SDK for C++ first.

File: exporting/aws_kinesis/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+To enable data sending to the kinesis service run `./edit-config exporting.conf` in the Netdata configuration directory
+and set the following options:
+
+```conf
+[kinesis:my_instance]
+    enabled = yes
+    destination = us-east-1
+```
+
+set the `destination` option to an AWS region.
```
Comment (shortpatti) [inline exporting/aws_kinesis/README.md]: Should "set" be "Set"?

File: exporting/aws_kinesis/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+## Prerequisites
+
+To use AWS Kinesis for metric collecting and processing, AWS SDK for C++ should be
+[installed](https://docs.aws.amazon.com/en_us/sdk-for-cpp/v1/developer-guide/setup.html) first. Netdata works with
+the SDK version 1.7.121. Other versions might work correctly as well but they were not tested with Netdata.
+`libcrypto`, `libssl`, and `libcurl` are also required to compile Netdata with Kinesis support enabled. Next, Netdata
+should be re-installed from the source. The installer will detect that the required libraries are now available.
+
+If the AWS SDK for C++ is being installed from source, it is useful to set `-DBUILD_ONLY="kinesis"`. Otherwise, the
+building process could take a very long time. Take a note, that the default installation path for the libraries is
```
Comment (joelhans) [inline exporting/aws_kinesis/README.md]:
```suggestion
building process could take a very long time. Note that the default installation path for the libraries is
```

File: exporting/aws_kinesis/README.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+`/usr/local/lib64`. Many Linux distributions don't include this path as the default one for a library search, so it is
+advisable to use the following options to `cmake` while building the AWS SDK:
+
+```sh
+cmake -DCMAKE_INSTALL_LIBDIR=/usr/lib -DCMAKE_INSTALL_INCLUDEDIR=/usr/include -DBUILD_SHARED_LIBS=OFF -DBUILD_ONLY=kinesis <aws-sdk-cpp sources>
+```
+
+## Configuration
+
+To enable data sending to the kinesis service run `./edit-config exporting.conf` in the Netdata configuration directory
```
Comment (joelhans) [inline exporting/aws_kinesis/README.md]:
```suggestion
To enable data sending to the Kinesis service, run `./edit-config exporting.conf` in the Netdata configuration directory
```

File: exporting/aws_kinesis/README.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+```conf
+[kinesis:my_instance]
+    enabled = yes
+    destination = us-east-1
+```
+
+set the `destination` option to an AWS region.
+
+Set AWS credentials and stream name:
```
Comment (joelhans) [inline exporting/aws_kinesis/README.md]: Where are you supposed to set this? In `exporting.conf`, or a different config file?

File: exporting/aws_kinesis/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+```yaml
+# AWS credentials
+aws_access_key_id = your_access_key_id
+aws_secret_access_key = your_secret_access_key
+
+# destination stream
+stream name = your_stream_name
+```
+
+Alternatively, AWS credentials can be set for the `netdata` user using AWS SDK for C++ [standard methods](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/credentials.html).
```
Comment (joelhans) [inline exporting/aws_kinesis/README.md]:
```suggestion
Alternatively, you can set AWS credentials for the `netdata` user using AWS SDK for C++ [standard methods](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/credentials.html).
```

File: exporting/aws_kinesis/README.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+A partition key for every record is computed automatically by Netdata with the purpose to distribute records across
```
Comment (joelhans) [inline exporting/aws_kinesis/README.md]:
```suggestion
Netdata automatically computes a partition key for every record with the purpose to distribute records across
```

File: exporting/read_config.c (hunk 2 above)
Comment (vlvkobal) [reply exporting/read_config.c]: Fixed

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/aws_kinesis/aws_kinesis_put_record.cc  2
- exporting/json/json.c  1
- exporting/tests/test_exporting_engine.c  8
- exporting/opentsdb/opentsdb.c  2
- exporting/aws_kinesis/aws_kinesis.c  3
- backends/aws_kinesis/aws_kinesis_put_record.cc  2
- backends/backends.c  2
         

Clones removed
==============
+ exporting/process_data.c  -16
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5042459&bid=16607224)

File: exporting/tests/exporting_doubles.c (hunk 3 above)
Comment (vlvkobal) [reply exporting/tests/exporting_doubles.c]: I don't think we need to handle all possible errors in tests. If the test crashes we can trace our code.

File: exporting/aws_kinesis/README.md (hunk 4 above)
Comment (vlvkobal) [reply exporting/aws_kinesis/README.md]: Done

File: exporting/aws_kinesis/README.md (hunk 8 above)
Comment (vlvkobal) [reply exporting/aws_kinesis/README.md]: Yes, in `exporting.conf`.

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/tests/test_exporting_engine.c  8
- exporting/json/json.c  1
- backends/backends.c  2
- exporting/opentsdb/opentsdb.c  2
- exporting/aws_kinesis/aws_kinesis.c  3
- backends/aws_kinesis/aws_kinesis_put_record.cc  2
- exporting/aws_kinesis/aws_kinesis_put_record.cc  2
         

Clones removed
==============
+ exporting/process_data.c  -16
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5042459&bid=16607224)

**Comment by stelfrag** [comment]
Ok got it to compile with the latest `aws-sdk-cpp` and making the change suggested in #8039 so it works with the latest version . 

The readme should be updated based on #8039 as well

**Comment by stelfrag** [comment]
Just noting this: 

In the configuration the aws access and secret keys are added with underscores
In general all keys in the config (also in netdata.conf) have spaces (there a few ones here and there that do have underscores though) 

_The aws cli config files expect those to have underscores in the ~/aws directory_


```
[kinesis:instance4]
enabled = yes
destination = us-east-1
stream name = netdata
aws_access_key_id = <id>
aws_secret_access_key = <secret>
data source = as collected
update every = 2
send charts matching = system.processes
```

**Comment by vlvkobal** [comment]
I'd not include anything from #8039. It's a separate issue and the solution is not that simple as to write a few lines in the readme. It also assumes changing the configuration of Autotools and CMake, checking for versions and dependencies. I think 1.7.121 is easier to install.

**Comment by vlvkobal** [comment]
`aws_access_key_id` and `aws_secret_access_key` are names borrowed from the AWS SDK configuration files.

**Comment by underhood** [comment]
OK so i do get it working with the AWS SDK version `1.7.121`. With the `1.7.281` i was not able to make it work on Fedora 31 AMD64. I get data back with  `aws kinesis get-records --shard-iterator=XXXX`. I will continue to review.

File: exporting/aws_kinesis/aws_kinesis.c [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+                connector_specific_config->stream_name,
+                partition_key,
+                buffer_len,
+                record_len);
+
+            kinesis_put_record(
+                connector_specific_data, connector_specific_config->stream_name, partition_key, first_char, record_len);
+
+            sent += record_len;
+            stats->chart_transmission_successes++;
```
Comment (stelfrag) [inline exporting/aws_kinesis/aws_kinesis.c]: Do you want to count this success if the `kinesis_get_result` indicates that there was a failure?

Comment (vlvkobal) [reply exporting/aws_kinesis/aws_kinesis.c]: No. It will be used for displaying internal statistics which is not implemented yet in the exporting engine.

**Comment by matthewmueller** [comment]
Newbie question: does this also work with AWS Firehose?

**Comment by vlvkobal** [comment]
No. Firehose is a separate service and has its own [client API](http://sdk.amazonaws.com/cpp/api/LATEST/namespace_aws_1_1_firehose.html).

**Pull request merged**